File: 20-valid-parentheses/20-valid-parentheses.go
// 20. Valid Parentheses
// time: O(n)
// space: O(n)

func isValid(s string) bool {
    var stack []byte
    for i := range s {
        switch ch := s[i]; ch {
        case '(':
            stack = append(stack, ')')
        case '[':
            stack = append(stack, ']')
        case '{':
            stack = append(stack, '}')
        case '}', ']', ')':
            if len(stack) == 0 {
                return false
            }
            if v := stack[len(stack)-1]; v != ch {
                return false
            }
            stack = stack[:len(stack)-1]
        }
    }
    return len(stack) == 0
}

File: 20-valid-parentheses/20-valid-parentheses.rs
// 20. Valid Parentheses
// time: O(n)
// space: O(n)

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = vec![];
        for ch in s.chars() {
            match ch {
                '(' => stack.push(')'),
                '[' => stack.push(']'),
                '{' => stack.push('}'),
                '}' | ']' | ')' if stack.pop() != Some(ch) => {
                    return false
                },
                _ => {},
            }
        }
    return stack.is_empty()
    }
}